toolchain/check/function.cpp

```diff
@@ -327,4 +327,42 @@ auto MergeFunctionRedecl(Context& context, SemIR::LocId loc_id,
   return true;
 }
 
+auto CheckFunctionReturnType(Context& context, SemIR::Function& function)
+    -> void {
+  // If we have already checked the return type, we have nothing to do.
+  if (function.return_slot != SemIR::Function::ReturnSlot::NotComputed) {
+    return;
+  }
+
+  if (!function.return_type_id.is_valid()) {
+    // Implicit `-> ()` has no return slot.
+    function.return_slot = SemIR::Function::ReturnSlot::Absent;
+    return;
+  }
+
+  // Check the return type is complete. Only diagnose incompleteness if we've
+  // not already done so.
+  auto diagnose_incomplete_return_type = [&] {
+    CARBON_DIAGNOSTIC(IncompleteTypeInFunctionReturnType, Error,
+                      "Function returns incomplete type `{0}`.", SemIR::TypeId);
+    return context.emitter().Build(function.return_storage_id,
+                                   IncompleteTypeInFunctionReturnType,
+                                   function.return_type_id);
+  };
+  std::optional<llvm::function_ref<auto()->Context::DiagnosticBuilder>>
```

Comment:
Would it be worth a typedef on Context for this?

Reply:
Rearranged to avoid the need for this type.